data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/s3/grouping/S3GroupIdentifier.java

```diff
@@ -43,6 +43,6 @@ public int hashCode() {
 
     public Map<String, Object> getGroupIdentifierHash() { return groupIdentifierHash; }
 
-    public Map<String, String> getMetadata(int eventCount) { return predefinedObjectMetadata != null ? Map.of(predefinedObjectMetadata.getNumberOfObjects(), Integer.toString(eventCount)) : null; }
+    public Map<String, String> getMetadata(int eventCount) { return objectMetadata != null ? Map.of(objectMetadata.getNumberOfEventsKey(), Integer.toString(eventCount)) : null; }
```

Comment:
This approach is coupling the metadata with the group itself. We should have a more extensible approach that allows for getting the metadata elsewhere.

Also, this design is intrinsically connected to the count. But, the metadata may be more than just this.

I think having a class to get metadata for any given S3 Object write would make more sense.

Reply:
We should have more extensible approach. We can do that in future. The current approached is NOT connected to count. `getMetadata()` returns a map which can be more than the count. Currently, it is just a count.

Reply:
S3 PutObjectRequest   API takes metadata as a MAP. We have ObjectMetadata class already being added in the PR. We can have API to populate it and return a MAP instead of creating map outside. I think this can be done in a future PR